## **PHASE 1 PROJECT**

## **Business Understanding**

### **Overview**
Skyway Airlines is a new airline that wants to enter the aviation industry. It wants to purchase aircrafts to operate for commercial and private enterprises. As a startup, one of the most important things is buying aircrafts that are safe and reliable for their operations. This project uses data from past aviation accidents between 1919 and 2023 to get potential risks of different aircrafts and determine the lowest risk aircraft. It also aims to show how important a specific airline's practices are having the best aircrafts or not. The goal is to provide Skyway Airlines with information to build a strong foundation of safety and efficiency for its operations.  

### **Problem Statement**
An airline company comes with high risks such as ensuring passenger safety and maintaining operational reliability. There are many issues that cause aviation accidents including the type of aircraft. Therefore, choosing the wrong type of aircraft would result in safety challenges and damage the brand reputation. Being unaware of the issues that cause accidents would also make it hard to ensure less accidents or little to no damage during accidents. All this may bring down the company as it is new in stead of growing it.

### **Objectives**
1. Determine the aircraft with the lowest risk to purchase for the business.
2. Find the potential risks of aircrafts and explain them to the stakeholders.
3. Determine how important airline practices are in ensuring safety regardless of the aircraft type.

### **Solution**
The project aims to guide Skyway Airlines to making informed decisions on purchasing aircrafts and on ensuring operational reliabilty. That is, it provides aircraft and operator safety insights. This is by analyzing past aviation accident data to identify potential risks and the safest aircraft type. The analysis also aims to show that a specific airline can be the cause of an accident even though the type of aircraft is the best.

## **Data Understanding** 

### Loading the data to a DataFrame, gettng its preview and metadata to understand the data.

In [1104]:
#Importing the pandas library for Explaratory Data Analysis
import pandas as pd

Loading the data to a pandas DataFrame from a csv file and viewing the DataFrame

In [1105]:
df = pd.read_csv('aviation-accident-data-2023-05-16.csv')
df

date                       type registration  \
0      date unk.             Antonov An-12B       T-1206   
1      date unk.             Antonov An-12B       T-1204   
2      date unk.             Antonov An-12B       T-1201   
3      date unk.            Antonov An-12BK          NaN   
4      date unk.            Antonov An-12BP   CCCP-11815   
...          ...                        ...          ...   
23962  11-May-23               Hawker 900XP       PK-LRU   
23963  11-May-23  Cessna 208B Grand Caravan       PK-NGA   
23964  12-May-23  Cessna 208B Grand Caravan       5X-RBR   
23965  14-May-23            Boeing 747-4R7F       LX-OCV   
23966  15-May-23                Learjet 35A       D-CGFQ   

                     operator fatalities             location  \
0               Indonesian AF        NaN                  NaN   
1               Indonesian AF        NaN                  NaN   
2               Indonesian AF        NaN                  NaN   
3                   Soviet AF        NaN  Tiksi Airport (IKS)   
4                   Soviet AF          0  Massawa Airport ...   
...                       ...        ...                  ...   
23962  Angkasa Super Services          0  Maleo Airport (MOH)   
23963  Nasional Global Aviasi          0    Fentheik Airstrip   
23964            Bar Aviation          0  Kampala-Kajjansi...   
23965                Cargolux          0  Luxembourg-Finde...   
23966                     GFD          2        Hohn Air Base   

               country cat     year  
0      Unknown country  U1  unknown  
1      Unknown country  U1  unknown  
2      Unknown country  U1  unknown  
3               Russia  A1  unknown  
4              Eritrea  A1  unknown  
...                ...  ..      ...  
23962        Indonesia  A2     2023  
23963        Indonesia  A2     2023  
23964           Uganda  A1     2023  
23965       Luxembourg  A2     2023  
23966          Germany  A1     2023  

[23967 rows x 9 columns]

Getting the first few rows of the data

In [1106]:
df.head(20)

date             type registration  \
0   date unk.   Antonov An-12B       T-1206   
1   date unk.   Antonov An-12B       T-1204   
2   date unk.   Antonov An-12B       T-1201   
3   date unk.  Antonov An-12BK          NaN   
4   date unk.  Antonov An-12BP   CCCP-11815   
5   date unk.  Antonov An-12BP   CCCP-12172   
6   date unk.     Antonov An-2    CCCP-N574   
7   date unk.     Antonov An-2   CCCP-01216   
8   date unk.   Antonov An-24B     RA-47794   
9   date unk.    Antonov An-26       01 red   
10  date unk.    Antonov An-26     RA-47415   
11  date unk.    Antonov An-26     RA-47414   
12  date unk.    Antonov An-26       5A-DOK   
13  date unk.    Antonov An-2R            -   
14  date unk.    Antonov An-2R       YU-BFR   
15  date unk.    Antonov An-2T   CCCP-01196   
16  date unk.    Antonov An-2T   CCCP-28909   
17  date unk.    Antonov An-2T   CCCP-98313   
18  date unk.    Antonov An-2T   CCCP-28895   
19  date unk.    Antonov An-2T   CCCP-35494   

                                operator fatalities             location  \
0                          Indonesian AF        NaN                  NaN   
1                          Indonesian AF        NaN                  NaN   
2                          Indonesian AF        NaN                  NaN   
3                              Soviet AF        NaN  Tiksi Airport (IKS)   
4                              Soviet AF          0  Massawa Airport ...   
5                              Soviet AF        NaN                  NaN   
6   GUSMP, Directorate of Polar Aviation        NaN              unknown   
7                               Aeroflot          0         Chita region   
8                             Russian AF          0                  NaN   
9                              Soviet AF          0    Orenburg Air Base   
10                            Russian AF          0  Belgorod Airport...   
11                              Aeroflot          0                  NaN   
12                      Sin Sad Airlines        NaN     Wadi Bu al Hashm   
13                            PZL-Mielec          0               Mielec   
14            Privredna Avijacija Osijek        NaN                  NaN   
15                  Aeroflot, Azerbaijan          0                  NaN   
16                  Aeroflot, Azerbaijan          0                  NaN   
17                  Aeroflot, Azerbaijan          0                  NaN   
18                  Aeroflot, Azerbaijan          0                  NaN   
19                  Aeroflot, Azerbaijan          0                  NaN   

            country cat     year  
0   Unknown country  U1  unknown  
1   Unknown country  U1  unknown  
2   Unknown country  U1  unknown  
3            Russia  A1  unknown  
4           Eritrea  A1  unknown  
5            Russia  U1  unknown  
6            Russia  A1  unknown  
7            Russia  A2  unknown  
8            Russia  A1  unknown  
9            Russia  O1  unknown  
10           Russia  A1  unknown  
11           Russia  A1  unknown  
12            Libya  A1  unknown  
13           Poland  O1  unknown  
14  Unknown country  U1  unknown  
15  Unknown country  O1  unknown  
16                ?  O1  unknown  
17                ?  O1  unknown  
18                ?  O1  unknown  
19                ?  O1  unknown

Getting the last few rows

In [1107]:
df.tail(15)

date                       type registration  \
23952  21-Apr-23            Antonov An-12BP            -   
23953  21-Apr-23              Antonov An-26         7705   
23954  21-Apr-23              Antonov An-26         7777   
23955  21-Apr-23              Antonov An-30         7708   
23956  21-Apr-23             Antonov An-32B         7709   
23957  21-Apr-23           Ilyushin Il-76TD       ST-BDN   
23958  21-Apr-23   Beech 200 Super King Air       LV-WPM   
23959  26-Apr-23           Boeing 777-233LR       C-FIUF   
23960   9-May-23     Cessna 208 Caravan 675       PK-HVG   
23961  10-May-23                Learjet 36A        N56PA   
23962  11-May-23               Hawker 900XP       PK-LRU   
23963  11-May-23  Cessna 208B Grand Caravan       PK-NGA   
23964  12-May-23  Cessna 208B Grand Caravan       5X-RBR   
23965  14-May-23            Boeing 747-4R7F       LX-OCV   
23966  15-May-23                Learjet 35A       D-CGFQ   

                     operator fatalities                  location  \
23952                Sudan AF          0       Khartoum Interna...   
23953                Sudan AF          0       Khartoum Interna...   
23954                Sudan AF          0       Khartoum Interna...   
23955                Sudan AF          0       Khartoum Interna...   
23956                Sudan AF          0       Khartoum Interna...   
23957           Badr Airlines          0       Khartoum Interna...   
23958               Flytec SA          0       Salta-Martín Mi...   
23959              Air Canada          0       Sydney-Kingsford...   
23960             Dimonim Air          0             Yabi Airstrip   
23961             Phoenix Air          3  near San Clemente Isl...   
23962  Angkasa Super Services          0       Maleo Airport (MOH)   
23963  Nasional Global Aviasi          0         Fentheik Airstrip   
23964            Bar Aviation          0       Kampala-Kajjansi...   
23965                Cargolux          0       Luxembourg-Finde...   
23966                     GFD          2             Hohn Air Base   

          country cat  year  
23952       Sudan  O1  2023  
23953       Sudan  O1  2023  
23954       Sudan  O1  2023  
23955       Sudan  O1  2023  
23956       Sudan  O1  2023  
23957       Sudan  O1  2023  
23958   Argentina  A2  2023  
23959   Australia  A2  2023  
23960   Indonesia  A2  2023  
23961         USA  A1  2023  
23962   Indonesia  A2  2023  
23963   Indonesia  A2  2023  
23964      Uganda  A1  2023  
23965  Luxembourg  A2  2023  
23966     Germany  A1  2023

Getting the metadata or information about the data, i.e How many entries and columns and what datatypes the data has for each column etc.

In [1108]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23967 entries, 0 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          23967 non-null  object
 1   type          23967 non-null  object
 2   registration  22419 non-null  object
 3   operator      23963 non-null  object
 4   fatalities    20029 non-null  object
 5   location      23019 non-null  object
 6   country       23967 non-null  object
 7   cat           23967 non-null  object
 8   year          23967 non-null  object
dtypes: object(9)
memory usage: 1.6+ MB


From the above, right now there are 23967 entries and 9 columns. There are missing values in the registration, operator, fatalities and location columns.   
date and year columns' datatype are string not datetime.  
Also, columns like date and year have values called 'date unk.' and 'unknown' respectively which should be taken as missing values.  
The same goes for the country column where 'Unknown country' and '?' are the same thing.


## **Data Cleaning**

#### **Checking for duplicates and dealing with them if any.**

In [1109]:
# How many duplicates are there
df.duplicated().sum()

115

In [1110]:
# Drop the duplicate rows
df = df.drop_duplicates()

Now the number of data entries have reduced hence get the metadata of the dataframe again.

In [1111]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23852 entries, 0 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          23852 non-null  object
 1   type          23852 non-null  object
 2   registration  22418 non-null  object
 3   operator      23848 non-null  object
 4   fatalities    20019 non-null  object
 5   location      22920 non-null  object
 6   country       23852 non-null  object
 7   cat           23852 non-null  object
 8   year          23852 non-null  object
dtypes: object(9)
memory usage: 1.8+ MB


The number of entries have reduced from 23967 to 23852.

#### **Dealing with missing values including the values that were unknown but have labels.**

Checking the number of missing values i.e original NaN values

In [1112]:
df.isna().sum()

date               0
type               0
registration    1434
operator           4
fatalities      3833
location         932
country            0
cat                0
year               0
dtype: int64

In [1113]:
# Importing the numpy library
import numpy as np

Replacing the values in date and year columns that were unknown but have labels with NaN to count as missing values.  
The unknown countries have no issue remaining 'Unknown country' since turning them to NaN creates too many missing values. Hence the '?' values in countries can also be 'Unknown country'.  
Also when trying to change date to datetime I noticed there are values with unknown day and/or month but have known year hence these values are also missing values.

In [1114]:
df.loc[:,'date'] = df['date'].replace('date unk.', np.nan)
df.loc[:,'date'] = df['date'].replace([r'^\?\?-\?\?\?-\d{4}$', r'^\?\?-[A-Z]{3}-\d{4}$', r'^\d{2}-\?\?\?-\d{4}$'], np.nan, regex=True)
df.loc[:,'year'] = df['year'].replace('unknown', np.nan)
df.loc[:,'country'] = df['country'].replace('?', 'Unknown country')

c:\Users\user\anaconda3\envs\learn-env\lib\site-packages\pandas\core\indexing.py:1783: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item_labels[indexer[info_axis]]] = value


Viewing the first 20 rows of the data again to see the difference made.

In [1115]:
df.head(20)

date             type registration                              operator  \
0   NaN   Antonov An-12B       T-1206                         Indonesian AF   
1   NaN   Antonov An-12B       T-1204                         Indonesian AF   
2   NaN   Antonov An-12B       T-1201                         Indonesian AF   
3   NaN  Antonov An-12BK          NaN                             Soviet AF   
4   NaN  Antonov An-12BP   CCCP-11815                             Soviet AF   
5   NaN  Antonov An-12BP   CCCP-12172                             Soviet AF   
6   NaN     Antonov An-2    CCCP-N574  GUSMP, Directorate of Polar Aviation   
7   NaN     Antonov An-2   CCCP-01216                              Aeroflot   
8   NaN   Antonov An-24B     RA-47794                            Russian AF   
9   NaN    Antonov An-26       01 red                             Soviet AF   
10  NaN    Antonov An-26     RA-47415                            Russian AF   
11  NaN    Antonov An-26     RA-47414                              Aeroflot   
12  NaN    Antonov An-26       5A-DOK                      Sin Sad Airlines   
13  NaN    Antonov An-2R            -                            PZL-Mielec   
14  NaN    Antonov An-2R       YU-BFR            Privredna Avijacija Osijek   
15  NaN    Antonov An-2T   CCCP-01196                  Aeroflot, Azerbaijan   
16  NaN    Antonov An-2T   CCCP-28909                  Aeroflot, Azerbaijan   
17  NaN    Antonov An-2T   CCCP-98313                  Aeroflot, Azerbaijan   
18  NaN    Antonov An-2T   CCCP-28895                  Aeroflot, Azerbaijan   
19  NaN    Antonov An-2T   CCCP-35494                  Aeroflot, Azerbaijan   

   fatalities             location          country cat year  
0         NaN                  NaN  Unknown country  U1  NaN  
1         NaN                  NaN  Unknown country  U1  NaN  
2         NaN                  NaN  Unknown country  U1  NaN  
3         NaN  Tiksi Airport (IKS)           Russia  A1  NaN  
4           0  Massawa Airport ...          Eritrea  A1  NaN  
5         NaN                  NaN           Russia  U1  NaN  
6         NaN              unknown           Russia  A1  NaN  
7           0         Chita region           Russia  A2  NaN  
8           0                  NaN           Russia  A1  NaN  
9           0    Orenburg Air Base           Russia  O1  NaN  
10          0  Belgorod Airport...           Russia  A1  NaN  
11          0                  NaN           Russia  A1  NaN  
12        NaN     Wadi Bu al Hashm            Libya  A1  NaN  
13          0               Mielec           Poland  O1  NaN  
14        NaN                  NaN  Unknown country  U1  NaN  
15          0                  NaN  Unknown country  O1  NaN  
16          0                  NaN  Unknown country  O1  NaN  
17          0                  NaN  Unknown country  O1  NaN  
18          0                  NaN  Unknown country  O1  NaN  
19          0                  NaN  Unknown country  O1  NaN

Checking the new number of missing values including the ones that had labels like 'unknown' for their values

In [1116]:
df.isna().sum()

date             548
type               0
registration    1434
operator           4
fatalities      3833
location         932
country            0
cat                0
year              61
dtype: int64

##### **Dropping some missing values**

Getting rid of rows with missing values except the ones where the 3 columns (type, fatalities, cat) all have values(not null). This is because these 3 columns are most important in getting insights and they are best used together.

In [1117]:
df = df.dropna(subset=['type', 'fatalities', 'cat'], how = 'any') # how='any' means if one of the 3 has a missing value, drop the column 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          19793 non-null  object
 1   type          20019 non-null  object
 2   registration  18755 non-null  object
 3   operator      20016 non-null  object
 4   fatalities    20019 non-null  object
 5   location      19853 non-null  object
 6   country       20019 non-null  object
 7   cat           20019 non-null  object
 8   year          19994 non-null  object
dtypes: object(9)
memory usage: 1.5+ MB


Getting a view of the first few rows remaining

In [1118]:
df.head()

date             type registration    operator fatalities  \
4   NaN  Antonov An-12BP   CCCP-11815   Soviet AF          0   
7   NaN     Antonov An-2   CCCP-01216    Aeroflot          0   
8   NaN   Antonov An-24B     RA-47794  Russian AF          0   
9   NaN    Antonov An-26       01 red   Soviet AF          0   
10  NaN    Antonov An-26     RA-47415  Russian AF          0   

               location  country cat year  
4   Massawa Airport ...  Eritrea  A1  NaN  
7          Chita region   Russia  A2  NaN  
8                   NaN   Russia  A1  NaN  
9     Orenburg Air Base   Russia  O1  NaN  
10  Belgorod Airport...   Russia  A1  NaN

##### **Filling the reamining missing values**

##### Filling the date missing values

Changing the date column from string to datetime and getting the metadata again to see this change.

In [1119]:
df['date'] = pd.to_datetime(df['date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          19793 non-null  datetime64[ns]
 1   type          20019 non-null  object        
 2   registration  18755 non-null  object        
 3   operator      20016 non-null  object        
 4   fatalities    20019 non-null  object        
 5   location      19853 non-null  object        
 6   country       20019 non-null  object        
 7   cat           20019 non-null  object        
 8   year          19994 non-null  object        
dtypes: datetime64[ns](1), object(8)
memory usage: 1.5+ MB


After dropping some missing values above, the first row is at index 4.  
Next I'm filling the first value of date column with 1st January 1919 since the values are ordered according to date and the first dates were of 1919.
This is so that I can fill the rest of the null date values with previous or next dates available.

In [1120]:
df.loc[4, 'date'] = pd.to_datetime("1919-01-01")

Filling date column missing values with the previous date or next date available since the data seems to be ordered according to the dates.  
Afterwards I'm getting the metadata to check the date non-null counts.  

In [1121]:
# df['date'] = df['date'].fillna(method='bfill')   # backward fill i.e fill with next date value
df['date'] = df['date'].ffill().bfill()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          20019 non-null  datetime64[ns]
 1   type          20019 non-null  object        
 2   registration  18755 non-null  object        
 3   operator      20016 non-null  object        
 4   fatalities    20019 non-null  object        
 5   location      19853 non-null  object        
 6   country       20019 non-null  object        
 7   cat           20019 non-null  object        
 8   year          19994 non-null  object        
dtypes: datetime64[ns](1), object(8)
memory usage: 2.2+ MB


Viewing the first 20 rows again to see the changes.

In [1122]:
df.head(20)

date                        type registration  \
4  1919-01-01             Antonov An-12BP   CCCP-11815   
7  1919-01-01                Antonov An-2   CCCP-01216   
8  1919-01-01              Antonov An-24B     RA-47794   
9  1919-01-01               Antonov An-26       01 red   
10 1919-01-01               Antonov An-26     RA-47415   
11 1919-01-01               Antonov An-26     RA-47414   
13 1919-01-01               Antonov An-2R            -   
15 1919-01-01               Antonov An-2T   CCCP-01196   
16 1919-01-01               Antonov An-2T   CCCP-28909   
17 1919-01-01               Antonov An-2T   CCCP-98313   
18 1919-01-01               Antonov An-2T   CCCP-28895   
19 1919-01-01               Antonov An-2T   CCCP-35494   
20 1919-01-01               Antonov An-32        K2688   
22 1919-01-01                     Avia 14       LZ-ILG   
25 1919-01-01              BN-2A Islander        N67HA   
35 1919-01-01            BN-2A-8 Islander       TZ-ASC   
40 1919-01-01        Douglas C-47A (DC-3)      HK-1505   
44 1919-01-01  Grumman G-159 Gulfstream I       5Y-JET   
46 1919-01-01             Ilyushin Il-14T   CCCP-04180   
49 1919-01-01             Ilyushin Il-76M   CCCP-86834   

                                operator fatalities             location  \
4                              Soviet AF          0  Massawa Airport ...   
7                               Aeroflot          0         Chita region   
8                             Russian AF          0                  NaN   
9                              Soviet AF          0    Orenburg Air Base   
10                            Russian AF          0  Belgorod Airport...   
11                              Aeroflot          0                  NaN   
13                            PZL-Mielec          0               Mielec   
15                  Aeroflot, Azerbaijan          0                  NaN   
16                  Aeroflot, Azerbaijan          0                  NaN   
17                  Aeroflot, Azerbaijan          0                  NaN   
18                  Aeroflot, Azerbaijan          0                  NaN   
19                  Aeroflot, Azerbaijan          0                  NaN   
20                             Indian AF          0  Kalaikunda Air F...   
22                      Balkan Bulgarian          0  Sofia-Vrazhdebna...   
25                       MD Air Services          0  Nassau Internati...   
35                               Unknown          0                  NaN   
40                               Unknown          0       within Bahamas   
44                                JetPet          0  Nairobi-Wilson A...   
46  GUSMP, Directorate of Polar Aviation          0  Ice Station Koms...   
49                             Soviet AF          0  Novokuznetsk Air...   

            country cat year  
4           Eritrea  A1  NaN  
7            Russia  A2  NaN  
8            Russia  A1  NaN  
9            Russia  O1  NaN  
10           Russia  A1  NaN  
11           Russia  A1  NaN  
13           Poland  O1  NaN  
15  Unknown country  O1  NaN  
16  Unknown country  O1  NaN  
17  Unknown country  O1  NaN  
18  Unknown country  O1  NaN  
19  Unknown country  O1  NaN  
20            India  A1  NaN  
22         Bulgaria  O1  NaN  
25          Bahamas  A1  NaN  
35  Unknown country  U1  NaN  
40          Bahamas  A1  NaN  
44            Kenya  A1  NaN  
46       Antarctica  A1  NaN  
49           Russia  A1  NaN

Viewing the last 15 rows to see the changes also.

In [1123]:
df.tail(15)

date                       type registration  \
23952 2023-04-21            Antonov An-12BP            -   
23953 2023-04-21              Antonov An-26         7705   
23954 2023-04-21              Antonov An-26         7777   
23955 2023-04-21              Antonov An-30         7708   
23956 2023-04-21             Antonov An-32B         7709   
23957 2023-04-21           Ilyushin Il-76TD       ST-BDN   
23958 2023-04-21   Beech 200 Super King Air       LV-WPM   
23959 2023-04-26           Boeing 777-233LR       C-FIUF   
23960 2023-05-09     Cessna 208 Caravan 675       PK-HVG   
23961 2023-05-10                Learjet 36A        N56PA   
23962 2023-05-11               Hawker 900XP       PK-LRU   
23963 2023-05-11  Cessna 208B Grand Caravan       PK-NGA   
23964 2023-05-12  Cessna 208B Grand Caravan       5X-RBR   
23965 2023-05-14            Boeing 747-4R7F       LX-OCV   
23966 2023-05-15                Learjet 35A       D-CGFQ   

                     operator fatalities                  location  \
23952                Sudan AF          0       Khartoum Interna...   
23953                Sudan AF          0       Khartoum Interna...   
23954                Sudan AF          0       Khartoum Interna...   
23955                Sudan AF          0       Khartoum Interna...   
23956                Sudan AF          0       Khartoum Interna...   
23957           Badr Airlines          0       Khartoum Interna...   
23958               Flytec SA          0       Salta-Martín Mi...   
23959              Air Canada          0       Sydney-Kingsford...   
23960             Dimonim Air          0             Yabi Airstrip   
23961             Phoenix Air          3  near San Clemente Isl...   
23962  Angkasa Super Services          0       Maleo Airport (MOH)   
23963  Nasional Global Aviasi          0         Fentheik Airstrip   
23964            Bar Aviation          0       Kampala-Kajjansi...   
23965                Cargolux          0       Luxembourg-Finde...   
23966                     GFD          2             Hohn Air Base   

          country cat  year  
23952       Sudan  O1  2023  
23953       Sudan  O1  2023  
23954       Sudan  O1  2023  
23955       Sudan  O1  2023  
23956       Sudan  O1  2023  
23957       Sudan  O1  2023  
23958   Argentina  A2  2023  
23959   Australia  A2  2023  
23960   Indonesia  A2  2023  
23961         USA  A1  2023  
23962   Indonesia  A2  2023  
23963   Indonesia  A2  2023  
23964      Uganda  A1  2023  
23965  Luxembourg  A2  2023  
23966     Germany  A1  2023

##### Filling the registration missing values
Filling the registration missing values with 'Unknown Registration' since it's a column of unique values hence I can't fill with any of the values available.  
Afterwards I'm getting the metadata to check the registration non-null counts.

In [1124]:
df['registration'] = df['registration'].fillna('Unknown Registration')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          20019 non-null  datetime64[ns]
 1   type          20019 non-null  object        
 2   registration  20019 non-null  object        
 3   operator      20016 non-null  object        
 4   fatalities    20019 non-null  object        
 5   location      19853 non-null  object        
 6   country       20019 non-null  object        
 7   cat           20019 non-null  object        
 8   year          19994 non-null  object        
dtypes: datetime64[ns](1), object(8)
memory usage: 2.2+ MB


##### Filling the operator missing values
Filling the operator missing values with 'Unknown Operator' since it's a column of specific values hence I can't fill with any of the values available.  
Afterwards I'm getting the metadata to check the operator non-null counts.

In [1125]:
df['operator'] = df['operator'].fillna('Unknown Operator')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          20019 non-null  datetime64[ns]
 1   type          20019 non-null  object        
 2   registration  20019 non-null  object        
 3   operator      20019 non-null  object        
 4   fatalities    20019 non-null  object        
 5   location      19853 non-null  object        
 6   country       20019 non-null  object        
 7   cat           20019 non-null  object        
 8   year          19994 non-null  object        
dtypes: datetime64[ns](1), object(8)
memory usage: 2.2+ MB


##### Filling the fatalities missing values

Changing the fatalities column from string type to integer type including the NaN values in the column

In [1126]:
df['fatalities'] = pd.to_numeric(df['fatalities'], errors='coerce').astype('Int64')
# 'I' supports nullable integers so the column can have integers and NaN values

Getting the value counts to know if the data is skewed or not hence to fill with median or mean respectively dependin on outcome.

In [1127]:
df['fatalities'].value_counts()

0      10705
2       1117
1       1005
3        968
4        873
       ...  
137        1
169        1
520        1
312        1
335        1
Name: fatalities, Length: 189, dtype: Int64

From the value counts, the data will be skewed to the left hence it's best to se the median to fill the null values.

In [1128]:
# Getting the median for the fatalities column
fatalities_median = df['fatalities'].median()
fatalities_median

0.0

In [1129]:
# fill fatalities missing values with the median
df['fatalities'] = df['fatalities'].fillna(fatalities_median)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          20019 non-null  datetime64[ns]
 1   type          20019 non-null  object        
 2   registration  20019 non-null  object        
 3   operator      20019 non-null  object        
 4   fatalities    20019 non-null  Int64         
 5   location      19853 non-null  object        
 6   country       20019 non-null  object        
 7   cat           20019 non-null  object        
 8   year          19994 non-null  object        
dtypes: Int64(1), datetime64[ns](1), object(7)
memory usage: 2.2+ MB


##### Filling the location missing values
Filling the location missing values with 'Unknown Location' since it's a column of specific values hence I can't fill with any of the values available.  
Afterwards I'm getting the metadata to check the location non-null counts.

In [1130]:
df['location'] = df['location'].fillna('Unknown Location')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          20019 non-null  datetime64[ns]
 1   type          20019 non-null  object        
 2   registration  20019 non-null  object        
 3   operator      20019 non-null  object        
 4   fatalities    20019 non-null  Int64         
 5   location      20019 non-null  object        
 6   country       20019 non-null  object        
 7   cat           20019 non-null  object        
 8   year          19994 non-null  object        
dtypes: Int64(1), datetime64[ns](1), object(7)
memory usage: 2.2+ MB


##### Filling the year missing values
Filling the year missing values with the year from the date column.  
Afterwards I'm getting the metadata to check the year non-null counts.

In [1131]:
df['year'] = df['year'].fillna(df['date'].dt.year.astype(str))
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20019 entries, 4 to 23966
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          20019 non-null  datetime64[ns]
 1   type          20019 non-null  object        
 2   registration  20019 non-null  object        
 3   operator      20019 non-null  object        
 4   fatalities    20019 non-null  Int64         
 5   location      20019 non-null  object        
 6   country       20019 non-null  object        
 7   cat           20019 non-null  object        
 8   year          20019 non-null  object        
dtypes: Int64(1), datetime64[ns](1), object(7)
memory usage: 2.2+ MB


##### **Confirming there are no duplicates or missing values**

In [1132]:
df.duplicated().sum()

0

In [1133]:
df.isna().sum()

date            0
type            0
registration    0
operator        0
fatalities      0
location        0
country         0
cat             0
year            0
dtype: int64

## **Data Analysis**